sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,555 @@
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Execution model
+
+The work-items in a SYCL ND-range kernel are arranged into a hierarchy of
+groups and associated scopes. Work-items can query their position within each
+group, and can coordinate with other members of the same group using group
+functions and group algorithms.
+
+The execution model for work-items in an ND-range kernel behaves as if the
+work-items are invoked in the following hypothetical implementation:
+
+- Execution of the entire ND-range kernel starts with a single root-group
+thread of execution.
+- The root-group thread of execution logically spawns one additional thread of
+execution for each work-group, and blocks with forward progress guarantee
+delegation upon those threads of execution.
+- Each of the work-group threads of execution logically spawns one additional
+thread of execution for each sub-group, and then the work-group thread of
+execution blocks with forward progress guarantee delegation upon
+those sub-group threads of execution.
+- Each of the sub-group threads of execution logically spawns an additional
+thread of execution for each work-item, and then the sub-group thread of
+execution blocks with forward progress guarantee delegation upon those
+work-item threads of execution.
+- Finally, each work-item thread of execution executes one instance of the
+kernel function.
+
+```mermaid
+graph TD;
+    subgraph root-group
+    root(Thread);
+    end
+    subgraph work-groups
+    root-->wg1(Thread);
+    root-->wg2(...);
+    end
+    subgraph sub-groups
+    wg1-->sg1(Thread);
+    wg1-->sg2(...);
+    end
+    subgraph work-items
+    sg1-->wi1(Thread);
+    sg1-->wi2(Thread);
+    sg1-->wi3(Thread);
+    sg1-->wi4(Thread);
+    sg1-->wi5(Thread);
+    sg1-->wi6(Thread);
+    sg1-->wi7(Thread);
+    sg1-->wi8(Thread);
+    end
+```
+
+[NOTE]
+====
+Blocking with forward progress guarantee delegation ensures that the forward
+progress guarantees provided by at least one thread of execution in a group
+are at least as strong as the forward progress guarantees provided by that
+group.
+====
+
+Safe coordination of work-items via memory requires careful attention to this
+hierarchy, since the ability of work-items to make progress depends on a
+combination of the forward progress guarantees provided by the work-item
+threads of execution and the forward progress guarantees provided by the
+sub-group, work-group and root-group threads of execution. Often, correct
+execution of a kernel requires knowledge of the forward progress guarantees
+of a certain subset of these threads of execution.
+
+To facilitate reasoning about work-item coordination, the forward progress
+guarantees of threads of execution in SYCL can be qualified by a _coordination
+scope_. This extension defines the following coordination scopes:
```

Comment:
I don't entirely know how to improve this, but the idea of a _coordination scope_ makes this complicated topic a bit more complicated and there is no C++ equivalent.  Can we maybe provide some informative intuition or advice for what to pass as a coordination scope in some different circumstances?

Reply:
I'm not sure if I've improved things, but I've taken a stab at this in 8836cfb55d03e857da27a0d6ff2603d8f847d92f.

Reply:
I think this is better, thanks!